src/Internal/Calculator/NativeCalculator.php

```diff
@@ -210,6 +210,30 @@ public function pow(string $a, int $e) : string
         return $result;
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    public function sqrt(string $a) : string
```

Comment:
Maybe we have to let the `$x / $ans` allow float string result?

I think it can solve this infinite loop problem.

Reply:
I don't think the other operators are bugged, they're well tested, the problem most likely lies in the sqrt algorithm. And no, you can't allow float, this would defeat the whole purpose of an arbitrary precision library, that must deal with *exact* calculations of any size.

Reply:
Well, I think it still about the numbers problem. Consider following code snippet:

```php
$ans = 3;
$x = 3;
$del = 0;
$pre = 0;
$count = 0;

while (abs($ans - $pre) > 0) {
    $count += 1;
    $pre = $ans;
    $ans = (int) ((int) ($ans +  (int) $x / $ans) / 2);
    var_dump($ans);
}
echo $count . PHP_EOL;
echo $ans;

```

It lets all `division` operator calculator will be the `integer`. And this will cause the infinite loop problem.

Reply:
AFAICS in this [Newton pseudocode](http://campus.murraystate.edu/academic/faculty/wlyle/420/Newton.htm), it looks like you're using a zero `Epsilon` (error tolerance), which returns the exact result but may end up with an infinite loop with some numbers in our case.

When I change `cmp($checkedSqrtNumber, '0')` to `cmp($checkedSqrtNumber, '1')`, the infinite loop is gone, but some tests fail, for example:

```
--- Expected
+++ Actual

-'7'
+'8'

-'66291728919896041699308382947726934'
+'66291728919896041699308382947726935'
```

Always wrong by 1 on the last digit. I guess this is where the issue lies, we need to use a zero epsilon indeed, but find a way to detect and prevent the infinite loop.